src/java/org/apache/cassandra/index/sai/disk/v1/IndexSearcher.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.SSTableQueryContext;
+import org.apache.cassandra.index.sai.disk.IndexSearcherContext;
+import org.apache.cassandra.index.sai.disk.PostingList;
+import org.apache.cassandra.index.sai.disk.PostingListRangeIterator;
+import org.apache.cassandra.index.sai.disk.PrimaryKeyMap;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.KeyRangeIterator;
+
+/**
+ * Abstract reader for individual segments of an on-disk index.
+ *
+ * Accepts shared resources (token/offset file readers), and uses them to perform lookups against on-disk data
+ * structures.
+ */
+public abstract class IndexSearcher implements Closeable
```

Comment:
Alternate naming ideas:

`IndexSearcher` -> `SegmentSearcher`
`InvertedIndexSearcher` -> `TrieSegmentSearcher`

I guess it's a question of whether you want "index" or "segment" to be implicit. Feels like "index" is more obvious, hence easier to leave out of the explicit name of the class.

Reply:
That might also mean `IndexSearcherContext` becomes `SegmentSearcherContext`

Reply:
If we are going to change it I would prefer `IndexSegmentSearcher` and `IndexSegmentSearcherContext`. The second is a bit verbose but it is clear in its meaning. I would also prefer for the instantiation to be `LiteralIndexSegmentSearcher`. Again, a bit verbose but clear in its meaning and follows the naming elsewhere.

Reply:
I'll take "a bit verbose but clear" any day of the week.